### DISCUSSION TITLE
MeshBuilder causes self intersections

### DISCUSSION THREAD

**Discussion by MughtyWinky** [mozman/ezdxf#611] [Q&A] [answered]
I'm trying to convert roof geometries, based on a 3d city model, into DXF. These roofs usually consist of >4 3d-points and should be exported as "areas".  Thus they are converted using the _MeshBuilder_. During conversion, the _MeshBuilder_ seems to call _subdivide_ngons_, which adds a point.  Depending on the geometry, this leads to self intersections. Is there a better way to solve this?

```python
import ezdxf

#create a new document
doc = ezdxf.new(dxfversion="R2013")
msp = doc.modelspace()
layout = doc.layouts.get("Model")

# create a easy 3d shape
# z = y for planarity
coords = [[0,0,0], [3,0,0], [3,1,1], [8,1,1], [8,0,0], [10,0,0], [10,10,10], [0,10,10], [0,0,0]]

# add coords as mesh
mb = ezdxf.render.MeshBuilder()
mb.add_face(coords)
# render calls subdivide_ngons,
# which adds a point at 4.67, 2.44
# and causes some self-intersections
mb.render_3dfaces(layout)

# how it should look
# but only as line
pl = msp.add_polyline3d(coords)
pl.rgb=[255,0,0]

doc.saveas("c:/temp/example.dxf")
```

![image](https://user-images.githubusercontent.com/26003212/149002450-3e92d9ee-4679-4ca3-be5b-41939c670346.png)

**Comment by mozman** [comment, marked as answer]
Ezdxf is not a sophisticated 3d library so try to keep things simple. In this case you add an n-gon to a `MeshBuilder` which is basically ok because the main target entity of `MeshBuilder` is the `MESH` entity, which support n-gons. The problem here is the rendering of the mesh as `3DFACE` objects, which supports only triangles and quadrilaterals but not n-gons. The subdiv algorithm which divides n-gons into triagles is very simple and can not handle concave shapes as you see. The only solution is to subdivide the n-gons into convex shapes by yourself or using the `MESH` entity. A third party library like [CGAL](https://github.com/Hui606/Install-CGAL-Bindings-for-python-and-java) may help to subdivide the polygon.

**Reply by MughtyWinky** [reply]
Thank's for the answer, that's what I assumed. Luckily the database backend _PostGIS_ provides _[ST_Tesselate()](https://postgis.net/docs/manual-3.2/ST_Tesselate.html)_, which uses _CGAL_ via _SFCGAL_. Something like
```SQL
SELECT 
	CASE 
		WHEN ST_NPoints(geom) >4 THEN ST_Tesselate(geom)
		ELSE geom
	END as geom
FROM
	test
```
will do the trick. But that's kind of offtopic.

I appreciate your quick answer and work at _ezdxf_. Thank you!

**Reply by mozman** [reply]
Added `MeshBuilder.tessellation()` method which uses the "ear clipping" algorithm to process ngons for rendering POLYFACE and 3DFACE entities. This should solve this issue.
